rpc/datapool.go

```diff
@@ -112,27 +119,140 @@ func (p *DataPool) GetCacheOrResolveBNS(deviceName string, client *Client) ([]Ad
+		if !p.bnsCacheUpdatingFlag[peerKey] && config.AppConfig.ResolveCacheTime > 0 {
+			if expireTime, ok := p.bnsCacheExpireItem[peerKey]; ok && time.Now().After(expireTime) {
+				p.bnsCacheUpdatingFlag[peerKey] = true
+				go p.updateCacheResolvePeers(deviceName, client)
+			}
+		}
+		return peers, nil
+	}
+
+	p.updateCacheResolvePeers(deviceName, client)
+
+	peers, cached = p.getCacheBNS(peerKey)
+	if cached {
+		return peers, nil
+	}
+
+	return nil, nil
+
+}
+
+// getCacheOrResolvePeersOfDrive get peers of drive address
+func (p *DataPool) GetCacheOrResolvePeersOfDrive(addr Address, client *Client) ([]Address, error) {
+
+	peerKey := fmt.Sprintf("peers:%s", addr.HexString())
+	peers, cached := p.getCacheBNS(peerKey)
+	if cached {
+		//check if cache is expired if so call updateCacheResolvePeersOfDrive async. Expire duration is config.AppConfig.ResolveCacheTime
+		if !p.bnsCacheUpdatingFlag[peerKey] && config.AppConfig.ResolveCacheTime > 0 {
+			if expireTime, ok := p.bnsCacheExpireItem[peerKey]; ok && time.Now().After(expireTime) {
+				p.bnsCacheUpdatingFlag[peerKey] = true
+				go p.updateCacheResolvePeersOfDrive(addr, client)
+			}
+		}
+		return peers, nil
+	}
+
+	p.updateCacheResolvePeersOfDrive(addr, client)
+
+	peers, cached = p.getCacheBNS(peerKey)
+	if cached {
+		return peers, nil
+	}
+
+	return nil, nil
+}
+
+// getCacheOrResolveMembersOfDriveMember get members of drive member address
+func (p *DataPool) GetCacheOrResolveMembersOfDriveMember(addr Address, client *Client) ([]Address, error) {
+	peerKey := fmt.Sprintf("members:%s", addr.HexString())
+	peers, cached := p.getCacheBNS(peerKey)
+	if cached {
+		//check if cache is expired if so call updateCacheResolveMembersOfDriveMember async. Expire duration is config.AppConfig.ResolveCacheTime
+		if !p.bnsCacheUpdatingFlag[peerKey] && config.AppConfig.ResolveCacheTime > 0 {
+			if expireTime, ok := p.bnsCacheExpireItem[peerKey]; ok && time.Now().After(expireTime) {
+				p.bnsCacheUpdatingFlag[peerKey] = true
+				go p.updateCacheResolveMembersOfDriveMember(addr, client)
+			}
+		}
+		return peers, nil
+	}
+
+	p.updateCacheResolveMembersOfDriveMember(addr, client)
+
+	peers, cached = p.getCacheBNS(peerKey)
 	if cached {
 		return peers, nil
 	}
 
+	return nil, nil
+}
+
+func (p *DataPool) updateCacheResolveMembersOfDriveMember(addr Address, client *Client) {
+	peerKey := fmt.Sprintf("members:%s", addr.HexString())
+	var peersList []Address
+	members, err := client.ResolveMembers(addr)
+	if err == nil {
+		peersList = append(peersList, members...)
+	}
+
+	p.SetCacheBNS(peerKey, peersList)
+	p.bnsCacheExpireItem[peerKey] = time.Now().Add(config.AppConfig.ResolveCacheTime)
+	p.bnsCacheUpdatingFlag[peerKey] = false
+}
+
+func (p *DataPool) updateCacheResolvePeersOfDrive(addr Address, client *Client) {
+	peerKey := fmt.Sprintf("peers:%s", addr.HexString())
+	var peersList []Address
+	members, err := client.ResolveMembers(addr)
+	if err == nil {
+		peersList = append(peersList, members...)
```

Comment:
If the members are smart contract addresses (e.g. DriveMember) then they should be excluded from the peerList